File: cpp/bob/bob.cpp
#include <algorithm>
#include "bob.h"

namespace bob {

    std::string trim(const std::string& s) {
        std::size_t start = 0;
        std::size_t end = s.length() - 1;
        for (; start < s.length() && isspace(s.at(start)); ++start);
        for (; end >= 0 && isspace(s.at(end)); --end);
        return s.substr(start, end + 1 - start);
    }

    bool is_upper(const std::string& message) {
        if (std::none_of(message.cbegin(), message.cend(), isalpha)) return false;
        for (const char ch : message) {
            if (isalpha(ch)) if (!isupper(ch)) return false;
        }
        return true;
    }

    std::string hey(const std::string& s) {
        if (std::all_of(s.cbegin(), s.cend(), isspace)) return "Fine. Be that way!";
        const std::string message = trim(s);
        if (message.at(message.length() - 1) == '?') {
            if (is_upper(message)) return "Calm down, I know what I'm doing!";
            return "Sure.";
        }
        if (is_upper(message)) return "Whoa, chill out!";
        return "Whatever.";
    }
}  // namespace bob

File: java/bob/src/main/java/Bob.java
/*

Since this exercise has a difficulty of > 4 it doesn't come
with any starter implementation.
This is so that you get to practice creating classes and methods
which is an important part of programming in Java.

Please remove this comment when submitting your solution.

*/
public class Bob {
    private static boolean isSpace(String s) {
        for (char c : s.toCharArray())
            if (!Character.isSpaceChar(c))
                return false;
        return true;
    }

    private static boolean isUpper(String s) {
        boolean atLeastOneLetter = false;
        for (char c : s.toCharArray())
            if (Character.isAlphabetic(c)) {
                atLeastOneLetter = true;
                if (!Character.isUpperCase(c))
                    return false;
            }
        return atLeastOneLetter;
    }

    public String hey(String text) {
        text = text.trim();
        if (isSpace(text)) return "Fine. Be that way!";
        if (isUpper(text)) return text.charAt(text.length() - 1) == '?' ? "Calm down, I know what I'm doing!" : "Whoa, chill out!";
        return text.charAt(text.length() - 1) == '?' ? "Sure." : "Whatever.";
    }
}